File: src/endpoints/login/post.ts
import { APIGatewayProxyHandler } from "aws-lambda";
import * as t from "io-ts";

import { getUser, IUser } from "src/dataEngine";
import { parseUnknown } from "src/utils";
import { compareHash } from "src/crypto";

const loginSchema = t.type({
  email: t.string,
  password: t.string
});

const handler: APIGatewayProxyHandler = async (event) => {
  let { errors: bodyErrors, value: body } = parseUnknown(loginSchema, event.body);
  if (bodyErrors.length > 0 || body === null) {
    return {
      statusCode: 400,
      body: JSON.stringify({
        error_code: "invalid_parameters"
      })
    };
  }

  const { errors: deErrors, value: user } = await getUser(body.email);
  if (deErrors.length > 0) {
    console.log("DE FAILURE");
    console.log(deErrors);
    return {
      statusCode: 500,
      body: JSON.stringify({
        error_code: "data_engine_failure"
      })
    };
  }

  if (user && compareHash(body.password, user.hash)) {
    return {
      statusCode: 200,
      body: JSON.stringify({
        name: user.name
      })
    };
  } else {
    return {
      statusCode: 401,
      body: JSON.stringify({
        error_code: "bad_credentials"
      })
    };
  }
};

export { handler };

File: api/cmd/login/post.go
package main

import (
	"encoding/json"
	"github.com/aws/aws-lambda-go/events"
	"github.com/aws/aws-lambda-go/lambda"
	"github.com/jbmanning/auth/api/internals/crypto"
	de "github.com/jbmanning/auth/api/internals/dataEngine"
	"github.com/jbmanning/auth/api/pkg/www"
)

type LoginParams struct {
	Email    string `json:"email"`
	Password string `json:"password"`
}

type LoginResponse struct {
	Errors []string `json:"errors"`
	Name   string   `json:"name"`
}

func Handler(request events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, error) {
	params := &LoginParams{}
	var user *de.User

	err := json.Unmarshal([]byte(request.Body), params)
	if err != nil {
		return www.GetResp(400, LoginResponse{
			Errors: []string{www.InvalidJSONError},
		})
	} else {
		user = de.GetUser(params.Email)
	}

	if user == nil || !crypto.CompareHashAndPassword(user.Hash, params.Password) {
		user = nil
		return www.GetResp(400, LoginResponse{
			Errors: []string{www.InvalidEmailPasswordError},
		})
	}

	code := 400
	if user != nil {
		code = 200
	}

	return www.GetResp(code, LoginResponse{
		Errors: []string{},
		Name:   user.Name,
	})
}

func main() {
	h := www.ErrorHandlingMiddleware(www.CORSMiddleware(Handler))
	lambda.Start(h)
}
